### PR TITLE
274: Add Deploy and Admin command 

### PR DESCRIPTION
**Pull request by diemonster**
**What does this pull request do?**

Adds the `deploy` and `admin` commands in the `command` package. 

**Todo**
- [x] Create
- [x] Read
- [x] List
- [x] Delete
- [x] Admin Debug

**How should this be tested?**

**Checklist**

[X] Manual tests
[X] Unit tests

### CODE REVIEW

File: cli/command/deploy.go [hunk 1]
```diff
@@ -1,15 +1,165 @@
...
+	for _, deploy := range deploys {
+		if name := deploy.DeployName; name != "" {
+			if _, exists := catalog[name]; !exists {
+				catalog[name] = deploy
+				continue
+			}
+
+			max, err := strconv.Atoi(catalog[name].Version)
+			if err != nil {
+				return nil, err
```
Comment (zpatrick) [inline cli/command/deploy.go]: let's wrap these errors so users can know which deploy is the culprite, e.g. 
```
return nil, fmt.Errorf("Deploy %s has an invalid version tag: %v", catalog[name].DeployID, err)
```

File: cli/command/deploy.go [hunk 2]
```diff
@@ -1,15 +1,165 @@
...
+			}
+
+			current, err := strconv.Atoi(deploy.Version)
+			if err != nil {
+				return nil, err
```
Comment (zpatrick) [inline cli/command/deploy.go]: ditto - `return nil, fmt.Errorf("Deploy %s has an invalid version tag: %v", deploy.DeployID, err)`

File: cli/command/deploy_test.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+	}
+
+	assert.Equal(t, len(output), 3)
+	// max 'a', 'b', and 'c' deploys
+	assert.Contains(t, output, input[2])
+	assert.Contains(t, output, input[5])
+	assert.Contains(t, output, input[8])
+}
+
+func createTempFile(t *testing.T, content string) (*os.File, func()) {
```
Comment (zpatrick) [inline cli/command/deploy_test.go]: I think the environment.Create PR is using this same helper function - we should probably put it in a common file

File: cli/command/admin.go [hunk 4]
```diff
@@ -10,6 +13,55 @@ func NewAdminCommand(b *CommandBase) *AdminCommand {
...
+		versionVerify = "disabled"
+	}
+
+	a.printer.Printf("DEBUG REPORT \n")
+	a.printer.Printf("------------ \n")
+	a.printer.Printf("Instance Name:    %v\n", debugInfo.Instance)
+	a.printer.Printf("API Endpoint:     %v\n", apiEndpoint)
+	a.printer.Printf("API Version:      %v\n", debugInfo.Version)
+	a.printer.Printf("CLI Version:      %v\n", cliVersion)
+	a.printer.Printf("CLI Auth:         %v\n", cliAuth)
```
[the remark this answers is not in the document]
Comment (diemonster) [reply cli/command/admin.go]: yeah I can remove. Previously seemed like there was value in remote debugging, but I think we have better options internally for that sort of thing.

File: cli/command/deploy.go [hunk 5]
```diff
@@ -1,15 +1,165 @@
...
+	})
+}
+
+func (d *DeployCommand) read(c *cli.Context) error {
+	args, err := extractArgs(c.Args(), "NAME")
+	if err != nil {
+		return err
+	}
+
+	deployIDs, err := d.resolver.Resolve("deploy", args["NAME"])
```
[the remark this answers is not in the document]
Comment (diemonster) [reply cli/command/deploy.go]: 😱  ty

File: cli/command/deploy_test.go (hunk 3 above)
Comment (diemonster) [reply cli/command/deploy_test.go]: Ahh cool, will move to `test_helpers.go`. I think the function rename should stick.

File: cli/command/deploy.go [hunk 6]
```diff
@@ -1,15 +1,169 @@
...
+		DeployName: args["NAME"],
+		DeployFile: content,
+	}
+
+	jobID, err := d.client.CreateDeploy(req)
+	if err != nil {
+		return err
+	}
+
+	return d.waitOnJobHelper(c, jobID, "creating", func(deployID string) error {
```
Comment (jparsons04) [inline cli/command/deploy.go]: [nit] I'm not sure how everyone else feels about this, but I've generally been [a little more verbose with the spinner text](https://github.com/quintilesims/layer0/blob/cli-refactor-loadbalancer-command/cli/command/load_balancer.go#L182). I'm curious about your thoughts about being more explicit about what the spinner shows.

File: cli/command/deploy.go [hunk 7]
```diff
@@ -1,15 +1,169 @@
...
+	return d.printer.PrintDeploys(deploys...)
+}
+
+func (d *DeployCommand) list(c *cli.Context) error {
+	deploySummaries, err := d.client.ListDeploys()
+	if err != nil {
+		return err
+	}
+
+	if !c.Bool("all") {
```
Comment (jparsons04) [inline cli/command/deploy.go]: I'm curious if this is going to cause unintended behavior given the provider is now only showing [active task definitions](https://github.com/quintilesims/layer0/blob/cli-refactor/api/provider/aws/deploy_list.go#L55) and [active task definition families](https://github.com/quintilesims/layer0/blob/cli-refactor/api/provider/aws/deploy_list.go#L38). Would a user expect to get a list of inactive task definitions if this flag is specified?

File: cli/command/deploy_test.go [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+
+	base.Client.EXPECT().
+		ReadDeploy("dpl_id").
+		Return(&models.Deploy{}, nil)
+
+	c := NewContext(t, []string{file.Name(), "dpl_name"}, nil)
+	if err := command.create(c); err != nil {
+		t.Fatal(err)
+	}
+}
```
Comment (jparsons04) [inline cli/command/deploy_test.go]: We should be folding in the functional patterns for testing the wait/no-wait config in this test.

File: cli/command/deploy_test.go [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+
+	base.Client.EXPECT().
+		ReadJob("jid").
+		Return(job, nil)
+
+	c := NewContext(t, []string{"dpl_name"}, nil)
+	if err := command.delete(c); err != nil {
+		t.Fatal(err)
+	}
+}
```
Comment (jparsons04) [inline cli/command/deploy_test.go]: Should we be using the [deleteHelper](https://github.com/quintilesims/layer0/blob/cli-refactor/cli/command/base.go#L46) in this function?

Comment (jparsons04) [reply cli/command/deploy_test.go]: To be a little more explicit, [this is how I'm using deleteHelper in my PR](https://github.com/quintilesims/layer0/blob/cli-refactor-loadbalancer-command/cli/command/load_balancer.go#L246)

File: cli/command/deploy.go (hunk 6 above)
Comment (diemonster) [reply cli/command/deploy.go]: We should have a consistent text style for all these `waitOnJobHelper()` calls --- most of them appear to be "creating" at present moment (or, at least, the ones I checked). I think we should decide on this outside of this PR discussion, though.

**Comment by zpatrick** [comment]
:+1: once the issues josh brought up are fixed

File: cli/command/deploy.go (hunk 6 above)
Comment (jparsons04) [reply cli/command/deploy.go]: :white_check_mark:

File: cli/command/deploy.go (hunk 7 above)
Comment (jparsons04) [reply cli/command/deploy.go]: :white_check_mark:

File: cli/command/deploy_test.go (hunk 9 above)
Comment (diemonster) [reply cli/command/deploy_test.go]: I am already using the delete helper in the actual code, this is in `deploy_test.go`

File: cli/command/deploy.go [hunk 10]
```diff
@@ -1,15 +1,169 @@
...
-	return cli.Command{}
+	return cli.Command{
+		Name:  "deploy",
+		Usage: "manage layer0 deploys",
+		Subcommands: []cli.Command{
+			{
+				Name:      "create",
+				Usage:     "create a new deploy",
+				Action:    d.create,
+				ArgsUsage: "PATH NAME",
```
Comment (jparsons04) [inline cli/command/deploy.go]: [nit] Based on the discussion about the ArgsUsage `NAME` becoming `DEPLOY_NAME`. I think this is the currently agreed consensus on the discussions I saw about this.

File: cli/command/deploy.go [hunk 11]
```diff
@@ -1,15 +1,169 @@
...
+			},
+			{
+				Name:      "delete",
+				Usage:     "delete a deploy",
+				ArgsUsage: "NAME",
```
Comment (jparsons04) [inline cli/command/deploy.go]: [nit] Same as above, `DEPLOY_NAME`

File: cli/command/deploy.go [hunk 12]
```diff
@@ -1,15 +1,169 @@
...
+				Action:    d.delete,
+			},
+			{
+				Name:      "get",
+				Usage:     "describe a deploy",
+				Action:    d.read,
+				ArgsUsage: "NAME",
```
Comment (jparsons04) [inline cli/command/deploy.go]: [nit] Same as above, `DEPLOY_NAME`

File: cli/command/deploy.go [hunk 13]
```diff
@@ -1,15 +1,169 @@
...
+						Usage: "list all versions of all deploys",
+					},
+				},
+			},
+		},
+	}
+}
+
+func (d *DeployCommand) create(c *cli.Context) error {
+	args, err := extractArgs(c.Args(), "PATH", "NAME")
```
Comment (jparsons04) [inline cli/command/deploy.go]: If you make the changes above, `NAME` must become `DEPLOY_NAME`

File: cli/command/deploy.go [hunk 14]
```diff
@@ -1,15 +1,169 @@
...
+		return err
+	}
+
+	content, err := ioutil.ReadFile(args["PATH"])
+	if err != nil {
+		return err
+	}
+
+	req := models.CreateDeployRequest{
+		DeployName: args["NAME"],
```
Comment (jparsons04) [inline cli/command/deploy.go]: If you make the changes above, `NAME` must become `DEPLOY_NAME`

File: cli/command/deploy.go [hunk 15]
```diff
@@ -1,15 +1,169 @@
...
+}
+
+func (d *DeployCommand) delete(c *cli.Context) error {
+	return d.deleteHelper(c, "deploy", func(deployID string) (string, error) {
+		return d.client.DeleteDeploy(deployID)
+	})
+}
+
+func (d *DeployCommand) read(c *cli.Context) error {
+	args, err := extractArgs(c.Args(), "NAME")
```
Comment (jparsons04) [inline cli/command/deploy.go]: If you make the changes above, `NAME` must become `DEPLOY_NAME`

File: cli/command/deploy.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jparsons04) [reply cli/command/deploy.go]: If you make the changes above, `NAME` must become `DEPLOY_NAME`

**Pull request merged**